Polygon aperture

egstern opened the pull request:
This PR adds the polygon aperture that defines an accepted region for transmitting particles defined by a set or vertices.

egstern committed:
start of PolygonAperture implementation

egstern committed:
PolygonAperture logic added and builds now.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,296 @@
+        : Named(std::move(name)),
+          Alignment(dx, dy, rotation_degree),
+          m_action(action), m_repeat_x(repeat_x), m_repeat_y(repeat_y), m_shift_odd_x(shift_odd_x),
+      vertices_x(vertices_xx), vertices_y(vertices_yy), min_radius2(min_radius2x)
```
There is a small subtlety missing here: You need to initialize the new object's `m_id` for book keeping, like this:
```suggestion
          m_action(action), m_repeat_x(repeat_x), m_repeat_y(repeat_y), m_shift_odd_x(shift_odd_x),
          m_id(PolygonApertureData::next_id),
          vertices_x(vertices_xx), vertices_y(vertices_yy), min_radius2(min_radius2x)
```
Example: https://github.com/BLAST-ImpactX/impactx/blob/25.10/src/elements/SoftSol.H#L174

We essentially use `m_id` to reference global variables outside of this class, in which we store _dynamic_ GPU data (i.e., vectors/arrays of runtime-determined length) of the instances `m_id`. This is primarily because we want to copy the whole of `PolygonAperture` to GPU in the current design, and the dynamic data will only be referenced by pointer for that purpose.

I am aware that this detail is a bit nifty right now, and I am not proud of it ;-) We can refactor the design pattern at some point, but it will likely become more verbose.

egstern committed:
Update src/elements/PolygonAperture.H

initializer for member object m_id

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

egstern committed:
Added input file parsing logic.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,297 @@
+         *  Particles outside the boundary are considered lost.
+         *
+         * @param action specify action of domain (transmit/absorb)
+         * @param vertices_x array of coordinates of the horizontal vertex positions (m)
+         * @param vertices_y array of coordinates of the vertical vertex positions (m)
```
```suggestion
         * @param vertices_xx array of coordinates of the horizontal vertex positions (m)
         * @param vertices_yy array of coordinates of the vertical vertex positions (m)
```
Doxygen doesn't like the fact that the two variable names given here are different from those that appear in the function argument list.

cemitch99 committed:
Update src/elements/PolygonAperture.H

egstern committed:
sense of test was incorrect

egstern committed:
input file example

egstern committed:
some problem merging upstream? I don't know what happened

egstern committed:
examples of aperture with offsets and rotations

cemitch99 commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,297 @@
+         * @param action specify action of domain (transmit/absorb)
+         * @param vertices_xx array of coordinates of the horizontal vertex positions (m)
+         * @param vertices_yy array of coordinates of the vertical vertex positions (m)
+         * @param min_radius2 a minimum radius**2 below which all particles are transmitted
```
```suggestion
         * @param min_radius2x a minimum radius**2 below which all particles are transmitted
```
Doxygen requires that the name in the documentation here coincides with the name of the function argument.

ax3l replied:
I think that was a typo below and is now fixed - the argument is called `min_radius2` indeed

egstern commented:
I can't seem to do that because I get warnings of shadowed variables.
That's why it is the way it is now. These are the warnings as I changed it
in branch egstern-polygon-apertures-xx.

````
n file included from
/home/egstern/src/impactx-test/src/elements/PolygonAperture.cpp:10:
/home/egstern/src/impactx-test/src/elements/PolygonAperture.H: In
constructor
'impactx::elements::PolygonAperture::PolygonAperture(std::vector<double>,
std::vector<double>, amrex::ParticleReal, amrex::ParticleReal,
amrex::ParticleReal, bool, Action, amrex::ParticleReal,
amrex::ParticleReal, amrex::ParticleReal,
std::optional<std::__cxx11::basic_string<char> >)':
/home/egstern/src/impactx-test/src/elements/PolygonAperture.H:97:33:
warning: declaration of 'min_radius2' shadows a member of
'impactx::elements::PolygonAperture' [-Wshadow]
   97 |             amrex::ParticleReal min_radius2,
      |             ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
/home/egstern/src/impactx-test/src/elements/PolygonAperture.H:285:29: note:
shadowed declaration is here
  285 |         amrex::ParticleReal min_radius2; // minimum radius in which
all pass squared
      |                             ^~~~~~~~~~~
/home/egstern/src/impactx-test/src/elements/PolygonAperture.H:96:47:
warning: declaration of 'vertices_y' shadows a member of
'impactx::elements::PolygonAperture' [-Wshadow]
   96 |             std::vector< amrex::ParticleReal> vertices_y,
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
/home/egstern/src/impactx-test/src/elements/PolygonAperture.H:284:43: note:
shadowed declaration is here
  284 |         std::vector< amrex::ParticleReal> vertices_y;
      |                                           ^~~~~~~~~~
/home/egstern/src/impactx-test/src/elements/PolygonAperture.H:95:47:
warning: declaration of 'vertices_x' shadows a member of
'impactx::elements::PolygonAperture' [-Wshadow]
   95 |             std::vector< amrex::ParticleReal> vertices_x,
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
/home/egstern/src/impactx-test/src/elements/PolygonAperture.H:283:43: note:
shadowed declaration is here
  283 |         std::vector< amrex::ParticleReal> vertices_x;
      |                                           ^~~~~~~~~~
```


On Tue, Nov 4, 2025 at 2:11 PM Chad Mitchell ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/elements/PolygonAperture.H
> <https://github.com/BLAST-ImpactX/impactx/pull/1199#discussion_r2491920856>
> :
>
> > +        {
> +            transmit,
> +            absorb
> +        };
> +
> +        static std::string
> +        action_name (Action const & action);
> +
> +        /** A thin collimator element that applies a transverse aperture boundary defined
> +         *  by points of a polygon vertices.
> +         *  Particles outside the boundary are considered lost.
> +         *
> +         * @param action specify action of domain (transmit/absorb)
> +         * @param vertices_xx array of coordinates of the horizontal vertex positions (m)
> +         * @param vertices_yy array of coordinates of the vertical vertex positions (m)
> +         * @param min_radius2 a minimum radius**2 below which all particles are transmitted
>
> ⬇️ Suggested change
>
> -         * @param min_radius2 a minimum radius**2 below which all particles are transmitted
> +         * @param min_radius2x a minimum radius**2 below which all particles are transmitted
>
> Doxygen requires that the name in the documentation here coincides with
> the name of the function argument.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BLAST-ImpactX/impactx/pull/1199#pullrequestreview-3418476778>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4V5DBWQCFFZQSKUD6RWAL33ECAFAVCNFSM6AAAAACKEJHM3GVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTIMJYGQ3TMNZXHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

egstern committed:
Rearrange storage of vertices data and min_radius2 so that
it should work on CPU and GPU and eliminate out-of-order
initialization warnings.

egstern committed:
script to plot the action of the aperture

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

egstern committed:
tweaks to plot layout

egstern committed:
tweaks to plot layout

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on examples/polygon_aperture/input_polygon_aperture.in:
```diff
@@ -0,0 +1,46 @@
+###############################################################################
```
Wonderful!

Please add your tests to `examples/CMakeLists.txt`

egstern committed:
New analysis script that checks tests the results of the aperture, small
changes to other scripts.

egstern committed:
I don't know why this changed upstream
Merge branch 'egstern-polygon-aperture' of github.com:egstern/impactx into egstern-polygon-aperture

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

egstern committed:
python bindings for polygon aperture

egstern committed:
python example file for the polygon aperture

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

egstern committed:
additional scripts demonstrating absorb, rotate, offset options

egstern committed:
add new polygon aperture tests to test suite

egstern committed:
changed upstream

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

egstern committed:
fix python stanza

egstern committed:
another fix python stanza

egstern committed:
changed upstream

egstern committed:
cleanup comments

egstern committed:
Fix type for binding of PolygonAperture class data member definitions.

cemitch99 commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,304 @@
+            // calculate whether point is inside polygon or not
+            T_Real r2 = powi<2>(u) + powi<2>(v);
+            // are we inside the minimum radius?
+            if (r2 < m_min_radius2) {
```
This appears to be the offending line when using SIMD, likely due to the fact that `r2`, as a variable of type `T_Real`, appears inside the conditional.  (Note that this was avoided in `Aperture.H`.)  We probably need to ask @ax3l or @WeiqunZhang to clarify the simplest correct usage with variables of type `T_Real` here.

ax3l replied:
Let's comment out the vectorize as I marked above for now. We can optimize the code in a follow-up

egstern committed:
Added the element to the examples documentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

egstern committed:
Added Python and Input-file documentation

egstern committed:
upstream changes

egstern committed:
spelling correction

ax3l commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,304 @@
+      public mixin::Thin,
+      public mixin::Alignment,
+      public mixin::NoFinalize,
+      public amrex::simd::Vectorized<amrex::simd::native_simd_size_particlereal>
```
```suggestion
      public mixin::NoFinalize
      // public amrex::simd::Vectorized<amrex::simd::native_simd_size_particlereal>
```

ax3l commented on examples/polygon_aperture/README.rst:
```diff
@@ -0,0 +1,92 @@
+
+       .. literalinclude:: input_fodo.in
+          :language: ini
+          :caption: You can copy this file from ``examples/polygon_aperture/input_fodo.in``.
```
```suggestion
          :caption: You can copy this file from ``examples/polygon_aperture/input_polygon_aperture.in``.
```

ax3l commented on examples/polygon_aperture/README.rst:
```diff
@@ -0,0 +1,92 @@
+      :Python:
+         ``examples/polygon_aperture/run_polygon_aperture_absorb_rotate_offset.py
+      :Input file:
+         ``examples/polygon_aperture/input_polygon_aperture_absorb_rotate_offset.in``
```
This part does not yet render well :)

ax3l commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,304 @@
+            T_Real v = (m_repeat_y == 0_prt) ? sy : (fmod(abs(sy)+dy, m_repeat_y)-dy);
+
+            // compare against the aperture boundary
+            VBool inside_aperture;
+
+            // calculate whether point is inside polygon or not
+            T_Real r2 = powi<2>(u) + powi<2>(v);
+            // are we inside the minimum radius?
+            if (r2 < m_min_radius2) {
+                inside_aperture = VBool(true); // yes!
+            } else{
+
+#if AMREX_DEVICE_COMPILE
+                amrex::ParticleReal* vertices_x = m_vertices_x_d_data;
+                amrex::ParticleReal* vertices_y = m_vertices_y_d_data;
+#else
+                amrex::ParticleReal* vertices_x = m_vertices_x_h_data;
+                amrex::ParticleReal* vertices_y = m_vertices_y_h_data;
+#endif
+
+                // no, we have to do the polygon calculation
+                T_Real thetasum = 0.0;
+                Complex v1conj, v2;
+
+                for (size_t i = 0; i < m_nvert-1; ++i) {
+                    v1conj = Complex(u - vertices_x[i], -(v-vertices_y[i]));
+                    v2 = Complex(u - vertices_x[i+1], v-vertices_y[i+1]);
+                    thetasum += arg(v2 * v1conj);
+                }
+                inside_aperture = abs(thetasum/(2*ablastr::constant::math::pi)) > 1.0e-12_prt;
```
That is a cool algorithm, do you like to add an in-line comment with a reference? :)

ax3l commented on examples/polygon_aperture/README.rst:
```diff
@@ -0,0 +1,92 @@
+.. _examples-fodo:
+
+Polygon Aperture
+=========
```
Just a warning in rst that we can fix:
```suggestion
Polygon Aperture
================
```

ax3l commented on examples/polygon_aperture/README.rst:
```diff
@@ -0,0 +1,92 @@
+      :language: python3
+      :caption: You can copy this file from ``examples/polygon_aperture/plot_polygon_aperture.py``.
+
+.. figure:: https://private-user-images.githubusercontent.com/16342668/513572482-f9029000-6327-45b2-bb10-5b5edbe60373.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI5ODc2NTcsIm5iZiI6MTc2Mjk4NzM1NywicGF0aCI6Ii8xNjM0MjY2OC81MTM1NzI0ODItZjkwMjkwMDAtNjMyNy00NWIyLWJiMTAtNWI1ZWRiZTYwMzczLnBuZz9YLUFtei1B ...
```
This needs to be a link to a public gist, I think

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -1049,6 +1049,20 @@ add_impactx_test(absorber.py
+    ON  # ImpactX MPI-parallel
+    examples/polygon_aperture/analysis_polygon_aperture.py
+    examples/polygon_aperture/plot_polygon_aperture.py
+)
```
@egstern do you like to add the other examples here, too?

You can skip the analysis by passing `OFF` for the parameters if analysis is not as easy as for the other test

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -527,6 +527,25 @@ This requires these additional parameters:
 
 
+``polygon_aperture``
+^^^^^^^^^^^^
```
```suggestion
``polygon_aperture``
^^^^^^^^^^^^^^^^^^^^
```

egstern committed:
Update examples/polygon_aperture/README.rst

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

egstern committed:
Update src/elements/PolygonAperture.H

Don't do simd here yet to avoid type error. Later, we'll redo this with an array of the proper bool type.

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

egstern committed:
Update examples/polygon_aperture/README.rst

ax3l: just a warning we can fix

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

egstern committed:
Update docs/source/usage/parameters.rst

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

egstern committed:
add comments describing the particle-in-polygon algorithm

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -527,6 +527,25 @@ This requires these additional parameters:
+^^^^^^^^^^^^^^^^^^^^
+
+``polygon_aperture`` for a thin collimator element applying a transverse polygon aperture boundary defined by :math:`(x,y)` coordinates
+and optional radius below which all particles are transmitted. The vertices must define a closed curve and be ordered in the counter-clockwise direction.
```
Closed curve: Below, you mention also that the last point provided by the user must be exactly the same as the first point. This is missing from the user-facing docs here, a user could assume that the last vertex automatically connects to the first.

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -1386,6 +1386,29 @@ This module provides elements and methods for the accelerator lattice.
+.. py:class:: impactx.elements.PolygonAperture(vertices_x, vertices_y, min_radius2=0.0, repeat_x, repeat_y, shift_odd_x, action="transmit", dx=0, dy=0, rotation=0, name=None)
+
+   This element defines a thin collimator element applying a transverse polygon aperture boundary defined by :math:`(x,y)` coordinates
+   and optional radius below which all particles are transmitted. The vertices must define a closed curve and be ordered in the counter-clockwise direction.
```
Closed curve: Below, you mention also that the last point provided by the user must be exactly the same as the first point. This is missing from the user-facing docs here, a user could assume that the last vertex automatically connects to the first.

ax3l commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,321 @@
+#include <AMReX_Math.H>
+#include <AMReX_REAL.H>
+#include <AMReX_SIMD.H>
+#include <ablastr/constant.H>
```
I need to clarify that ABLASTR goes before AMReX [includes](https://impactx.readthedocs.io/en/latest/developers/repo_organization.html#c-includes), because it depends on the latter.

```suggestion
#include <ablastr/constant.H>

#include <AMReX_Extension.H>
#include <AMReX_Math.H>
#include <AMReX_REAL.H>
#include <AMReX_SIMD.H>

```

ax3l commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,321 @@
+#include <AMReX_SIMD.H>
+#include <ablastr/constant.H>
+#include <cmath>
+#include <stdexcept>
```
```suggestion
#include <map>
#include <stdexcept>
#include <vector>
```

ax3l commented on src/elements/PolygonAperture.H:
```diff
@@ -0,0 +1,321 @@
+            // there must be the same number of horizontal and vertical vertices
+            if (m_nvert != vertices_y.size()) {
+                throw std::runtime_error("PolygonAperture: horizontal and vertical vertices arrays must have same length!");
+            }
```
Can you please add a runtime check, comparing that the first and last arguments of each vertix are exactly the same, as required by your algorithm?

egstern replied:
resolved in commit fbfce19136d76af566a655fadca2dc6c122d7f3d

ax3l reviewed (commented):
Looks great, Eric!

I added a review with inline comments now :)

egstern committed:
Update src/elements/PolygonAperture.H

ax3l: I need to clarify that ABLASTR goes before AMReX includes, because it depends on the latter.

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

egstern committed:
Update src/elements/PolygonAperture.H

from ax3l

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

egstern committed:
update documentation to include additional examples

egstern committed:
add test that first and last vertices are equal

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

egstern committed:
Clarify that the first and last vertices must be identical.

egstern committed:
Add additional example scripts to testing list

egstern committed:
upstream changes
Merge branch 'egstern-polygon-aperture' of github.com:egstern/impactx into egstern-polygon-aperture

egstern committed:
fix indent and close unterminated literal string to squash sphinx warnings

ax3l commented on examples/polygon_aperture/README.rst:
```diff
@@ -0,0 +1,96 @@
+      :language: python3
+      :caption: You can copy this file from ``examples/polygon_aperture/plot_polygon_aperture.py``.
+
+.. figure:: https://private-user-images.githubusercontent.com/16342668/513572482-f9029000-6327-45b2-bb10-5b5edbe60373.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI5ODc2NTcsIm5iZiI6MTc2Mjk4NzM1NywicGF0aCI6Ii8xNjM0MjY2OC81MTM1NzI0ODItZjkwMjkwMDAtNjMyNy00NWIyLWJiMTAtNWI1ZWRiZTYwMzczLnBuZz9YLUFtei1B ...
```
```suggestion
.. figure:: https://private-user-images.githubusercontent.com/16342668/513572482-f9029000-6327-45b2-bb10-5b5edbe60373.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM2NzQ4MTcsIm5iZiI6MTc2MzY3NDUxNywicGF0aCI6Ii8xNjM0MjY2OC81MTM1NzI0ODItZjkwMjkwMDAtNjMyNy00NWIyLWJiMTAtNWI1ZWRiZTYwMzczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTIwVDIxMzUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0NmMwMDI2MmI0ZTNlYzQyYmIxNWVmZTY0M2RiMjlmMWQyOTE3NzU2ZWE3OGRiYWQyM2ViMjE4NDEyMDgxNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ogbVVjFdQOVnTFkCgWXL72VyrhCnfEngwS3u5Pr1t-M
```

ax3l replied:
But feel free to  check this 60kb file in

ax3l replied:
20KB in the end, easy :)

egstern committed:
Merge branch 'development' into egstern-polygon-aperture

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

egstern committed:
Check in example plot file and point to it in the documentation.

cemitch99 reviewed (approved):
Looks good!

ax3l reviewed (approved):
This looks great, thanks a lot, Eric! Happy Thanksgiving :turkey:
